src/js/common/lib/i18n.js

```diff
@@ -0,0 +1,188 @@
+        if (candidateLocale === "*") {
+            return defaultLocale;
+        }
+        else if (localesAndLanguages.indexOf(candidateLocale) !== -1) {
+            return candidateLocale;
+        }
+    });
+
+    var preferredLanguageOrLocale = directMatchingLocaleOrLanguage || inheritedMatchingLanguage || defaultLocale;
+    return gpii.handlebars.i18n.deriveMessageBundle(preferredLanguageOrLocale, messageBundles, defaultLocale);
+};
+
+/**
+ *
+ * Derive a single message bundle from all available languages and locales, based on the `Accept-Language` headers
+ * passed as part of `request`.  Returns a combination of all messages found matching the following:
+ *
+ * 1. The user's locale.
+ * 2. The user's language.
+ * 3. The default locale.
+ *
+ * @param {String} preferredLocale - The locale to use.
+ * @param {Object} messageBundles - The full set of all message bundles available, keyed by locale or language.
+ * @param {String} defaultLocale - The default locale.  Compared with `locale` to determine how material should be inherited from the defaults.
+ * @return {Object} - A map of messages for a single locale.
+ *
+ */
+gpii.handlebars.i18n.deriveMessageBundle = function (preferredLocale, messageBundles, defaultLocale) {
+    defaultLocale = defaultLocale || "en_us";
+    preferredLocale = preferredLocale || defaultLocale;
+
+    var defaultMessages = fluid.get(messageBundles, defaultLocale);
+
+    var languageMessages = {};
+    // Look up any messages inherited from the language portion of the locale, and use those instead of the defaults.
+    var preferredLanguage = gpii.handlebars.i18n.languageFromLocale(preferredLocale);
+    if (preferredLanguage && fluid.get(messageBundles, preferredLanguage)) {
+        languageMessages = messageBundles[preferredLanguage];
+    }
+
+    var combinedMessageBundle = fluid.copy(languageMessages);
+
+    // The default locale wins out over messages inherited from the underlying language.
+    if (preferredLocale === defaultLocale) {
+        combinedMessageBundle = fluid.merge({}, combinedMessageBundle, defaultMessages);
+    }
+    // If we are not using the default locale and there are messages for our locale, those win over anything we've collected so far.
+    else if (fluid.get(messageBundles, preferredLocale)) {
+        combinedMessageBundle = fluid.merge({}, defaultMessages, combinedMessageBundle, messageBundles[preferredLocale]);
+    }
+
+    return combinedMessageBundle;
+};
+
+/**
+ *
+ * A sorting function which we use to prefer locales over languages.
+ *
+ * @param {String} a - A string representing a language or locale.
+ * @param {String} b - Another string representing a language or locale.
+ * @return {Integer} - A positive number if the second item is longer, a negative number if the first item is longer, or zero if the items are of the same length.
+ *
+ */
+gpii.handlebars.i18n.sortByDescendingLength = function (a, b) {
```

Comment:
Function is untested - misnamed at line 167

Reply:
Fixed.